File: backend/src/main/java/no/ntnu/idatt2106/krisefikser/service/InventoryService.java
package no.ntnu.idatt2106.krisefikser.service;

import no.ntnu.idatt2106.krisefikser.dto.HouseholdItemRequest;
import no.ntnu.idatt2106.krisefikser.dto.HouseholdItemResponse;
import no.ntnu.idatt2106.krisefikser.dto.UpsertInventoryRequest;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for managing household inventory items.
 * Provides methods to list, add, update, and remove items from a household's inventory.
 */
public interface InventoryService {
    List<HouseholdItemResponse> list(Integer householdId);
    HouseholdItemResponse upsert(Integer householdId, UpsertInventoryRequest req);
    HouseholdItemResponse add(Integer householdId, HouseholdItemRequest req);
    HouseholdItemResponse update(Integer householdId, Integer itemId, HouseholdItemRequest req);
    void remove(Integer householdId, Integer itemId, LocalDate acquiredDate);
    void removeAll(Integer householdId, Integer itemId);
}

File: frontend/src/api/InventoryService.ts
import axios from 'axios';

import type {
    HouseholdItemRequest,
    HouseholdItemResponse,
    UpsertInventoryRequest,
} from '@/types/Inventory';

const HOUSEHOLD_API_URL = 'http://localhost:8080/api/households/items';

export class InventoryService {
    static async list(): Promise<HouseholdItemResponse[]> {
        const res = await axios.get(`${HOUSEHOLD_API_URL}`, { withCredentials: true });
        return res.data;
    }

    static async add(item: HouseholdItemRequest): Promise<HouseholdItemResponse> {
        const res = await axios.post(`${HOUSEHOLD_API_URL}`, item, { withCredentials: true });
        return res.data;
    }

    static async update(itemId: number, item: HouseholdItemRequest): Promise<HouseholdItemResponse> {
        const res = await axios.put(`${HOUSEHOLD_API_URL}/${itemId}/${item.acquiredDate}`, item, { withCredentials: true });
        return res.data;
    }
    
    static async upsert(item: UpsertInventoryRequest): Promise<HouseholdItemResponse> {
        const res = await axios.post(`${HOUSEHOLD_API_URL}/upsert`, item, { withCredentials: true });
        return res.data;
    }
    
    static async remove(itemId: number, acquiredDate: string): Promise<void> {
        await axios.delete(`${HOUSEHOLD_API_URL}/${itemId}/${acquiredDate}`, { withCredentials: true });
    }
}